components/campaign/CampaignChaosBagView/index.tsx

```diff
@@ -0,0 +1,382 @@
+        }],
+      },
+    };
+  }
+
+  _navEventListener?: EventSubscription;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      chaosBagResults: {
+        currentToken: undefined,
+        drawnTokens: [],
+        total: 0,
+      },
+      visible: true,
+    };
+
+    this._navEventListener = Navigation.events().bindComponent(this);
+  }
+
+  componentDidMount() {
+    BackHandler.addEventListener('hardwareBackPress', this._handleBackPress);
+  }
+
+  componentWillUnmount() {
+    BackHandler.removeEventListener('hardwareBackPress', this._handleBackPress);
+    this._navEventListener && this._navEventListener.remove();
+  }
+
+  componentDidAppear() {
+    this.setState({
+      visible: true,
+    });
+  }
+
+  componentDidDisappear() {
+    this.setState({
+      visible: false,
+    });
+  }
+
+  navigationButtonPressed({ buttonId }: { buttonId: string }) {
+    if (buttonId === 'back' || buttonId === 'androidBack') {
+      this._handleBackPress();
+    } else if (buttonId === 'edit') {
+      this._showChaosBagDialog();
+    }
+  }
+
+  _handleDrawTokenPressed = () => {
+    this._drawToken();
+  }
+
+  _handleSealTokenPressed = () => {
+
+  }
+
+  _handleBackPress = () => {
+    const {
+      componentId,
+    } = this.props;
+    const {
+      visible,
+    } = this.state;
+    if (!visible) {
+      return false;
+    }
+    Navigation.pop(componentId);
+    return true;
+  }
+
+  _showChaosBagDialog = () => {
+    const {
+      componentId,
+      updateChaosBag,
+      chaosBag,
+    } = this.props;
+
+    Navigation.push<EditChaosBagProps>(componentId, {
+      component: {
+        name: 'Dialog.EditChaosBag',
+        passProps: {
+          chaosBag,
+          updateChaosBag: updateChaosBag,
+          trackDeltas: true,
+        },
+        options: {
+          topBar: {
+            title: {
+              text: t`Chaos Bag`,
+            },
+            backButton: {
+              title: t`Cancel`,
+            },
+          },
+        },
+      },
+    });
+  }
+
+  _getRandomNumber = (min: number, max: number) => {
+    return Math.random() * (max - min) + min;
+  }
+
+  _getRandomChaosToken = (chaosBag: ChaosBag) => {
+    const list = keys(chaosBag);
+    const weight = values(chaosBag);
+
+    const totalWeight = weight.reduce((prev, cur) => {
+      return (prev && cur) ? (prev + cur) : 0;
+    });
+
+    if (totalWeight) {
+      const random_num = this._getRandomNumber(0, totalWeight);
+      let weight_sum = 0;
+      const length = list.length;
+      for (let i = 0; i < length; i++) {
+        const item = weight[i];
+        if (item) {
+          weight_sum += item;
+          weight_sum = +weight_sum.toFixed(2);
+        }
+
+        if (random_num <= weight_sum) {
+          const tokenName = list[i];
+          let drawnToken = null;
+          CHAOS_TOKENS.forEach(token => {
+            if (token === tokenName) {
+              drawnToken = token;
```

Comment:
You could just do the return here directly -- also I believe CHAOS_TOKENS can be indexed into (or if not there should be a map that can be nearby it).